resource-providers/src/main/java/io/opentelemetry/resourceproviders/ParseBuddy.java

```diff
@@ -143,41 +147,53 @@ private static DescriptorHandler newAppXmlHandler() {
   private static final class DescriptorHandler extends DefaultHandler {
     private final String rootElementName;
     private final Deque<String> currentElement = new ArrayDeque<>();
-    private boolean setDisplayName;
-    @Nullable private String displayName;
+    @Nullable private String key;
+    private final Map<String, String> names = new HashMap<>();
 
     DescriptorHandler(String rootElementName) {
       this.rootElementName = rootElementName;
     }
 
     @Override
     public void startElement(String uri, String localName, String qName, Attributes attributes) {
-      if (displayName == null
-          && rootElementName.equals(currentElement.peek())
-          && "display-name".equals(qName)) {
-        String lang = attributes.getValue("xml:lang");
-        if (lang == null || "".equals(lang)) {
-          lang = "en"; // en is the default language
-        }
-        if ("en".equals(lang)) {
-          setDisplayName = true;
+      List<String> nameElements = Arrays.asList("display-name", "servlet-name");
+      if ((names.size() < 2) // If we haven't already found both names
+          && rootElementName.equals(
```

Comment:
This feels a bit confusing to me. Previous check `rootElementName.equals(currentElement.peek())` verified that currently handled `<display-name>` element is a child of the root element. Current check `rootElementName.equals(currentElement.peekLast())` verifies that the document has the expected root element and accepts `display-name` in any position. If the desire is to just verify the root element you don't need the current element stack at all. `<display-name>` can appear in multiple positions in web.xml. It can appear as a child of the root `<web-app>` or under `<servlet>` or `<filter>` maybe more. Are you sure it makes sense to use `<display-name>` for servlet or filter, they are likely to meaningless, something like `Main servlet` same for the `<servlet-name>`?

Reply:
I added a test case and tweaked the data to make it a little clearer that the outer `<display-name>` is always preferred. @laurit lemme know what you think.

Reply:
I think the main question here is whether it is preferable to detect a bad name or no name at all. If your goal is to detect some kind of name no matter what then I think you should consider doing multiple passes over the descriptors. In first pass consider only the top level `<display-name>` and if that doesn't yield a result in any of the deployed apps try something else. Perhaps detecting name from different sources should be structured as different resource providers so they could be enabled/disabled separately? I suspect that using `<servlet-name>` won't always work well will as it yields the same name for different apps because the web.xml for newer app was copied from an older app and the main servlet has the same name. Maybe it would be more worthwhile to look for alternate sources for service name like maven metadata.